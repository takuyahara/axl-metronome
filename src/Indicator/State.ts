enum State {
  Running = "Running",
  Stopped = "Stopped",
  Paused = "Paused",
  Dragging = "Dragging",
}
export default State;