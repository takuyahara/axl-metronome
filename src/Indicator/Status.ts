enum Status {
  Running = "Running",
  Stopped = "Stopped",
  Paused = "Paused",
}
export default Status;