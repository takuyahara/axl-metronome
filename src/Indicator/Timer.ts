import State from './State';

const DECIMAL = 5;

interface IPostProcess {
  onCount: (remaining: number) => void;
  onCountStart: () => void;
  onCountStop: () => void;
  onCountComplete: (remaining: number) => void;
}
class Timer {
  // Post process
  private postCount!: IPostProcess['onCount'];
  private postCountStart!: IPostProcess['onCountStart'];
  private postCountStop!: IPostProcess['onCountStop'];
  private postCountComplete!: IPostProcess['onCountComplete'];

  // Initialize in init()
  private state!: State;
  private timeToCount!: number;
  private timerId!: NodeJS.Timer | number;
  private remaining!: number;

  public constructor(timeToCount: number, postProcess: IPostProcess) {
    this.init(timeToCount, postProcess);
    this.timerRunner = this.timerRunner.bind(this);
    // this.postCount = this.postCount.bind(this);
    // this.postCountStart = this.postCountStart.bind(this);
    // this.postCountStop = this.postCountStop.bind(this);
    // this.postCountComplete = this.postCountComplete.bind(this);
  }
  private init(timeToCount: number, postProcess: IPostProcess): void {
    this.postCount = postProcess.onCount;
    this.postCountStart = postProcess.onCountStart;
    this.postCountStop = postProcess.onCountStop;
    this.postCountComplete = postProcess.onCountComplete;
    this.state = State.Stopped;
    this.timeToCount = timeToCount;
    this.remaining = timeToCount;
  }
  private isTimerCompleted(): boolean {
    return this.remaining < 0;
  }
  private fixProgress(progressRaw: number): number {
    // progressRaw can be 0.19999999999999996
    // fix to 0.2 (0.20000 when DECIMAL = 5)
    return Math.round(progressRaw * Math.pow(10, DECIMAL)) / Math.pow(10, DECIMAL);
  }
  private timerRunner(): void {
    this.postCount(--this.remaining);
    if (this.isTimerCompleted()) {
      this.reset();
      this.postCountComplete(this.remaining);
    }
  }
  private start(): void {
    this.timerId = setInterval(this.timerRunner, 1000);
    this.postCountStart();
    this.state = State.Running;
  }
  public stop(): void {
    clearInterval(this.timerId as NodeJS.Timer);
    this.remaining = this.timeToCount;
    this.postCountStop();
    this.state = State.Stopped;
  }
  public pause(): void {
    clearInterval(this.timerId as NodeJS.Timer);
    this.state = State.Paused;
  }
  public resume(): void {
    this.start();
  }
  public drag(): void {
    this.pause();
    this.state = State.Dragging;
  }
  public reset(): void {
    this.stop();
  }
  public getState(): State {
    return this.state;
  }
  public setRemaining(remaining: number): void {
    this.remaining = remaining;
  }
  public getRemaining(): number {
    return this.remaining;
  }
  public setTimeToCount(timeToCount: number): void {
    this.timeToCount = timeToCount;
    this.remaining = timeToCount;
  }
  public getTimeToCount(): number {
    return this.timeToCount;
  }
  public setProgress(progress: number): void {
    this.remaining = this.timeToCount * (1 - progress);
  }
  public getProgress(): number {
    const progressRaw = 1 - (this.remaining / this.timeToCount); // can be 0.19999999999999996
    const progressFixed = Math.round(progressRaw * Math.pow(10, DECIMAL)) / Math.pow(10, DECIMAL); // fix to 0.2 (0.20000 when DECIMAL = 5)
    return progressFixed;
  }
  public toggle(): void {
    switch (this.state) {
        case State.Running:
        this.pause();
        break;
      case State.Paused:
        this.resume();
        break;
      case State.Stopped:
        this.start();
        break;
    }
  }
}
export default Timer;
