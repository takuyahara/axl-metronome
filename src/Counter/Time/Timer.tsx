import React, { createRef, RefObject } from 'react';
import _ from 'lodash';
import Utils from '../../Utils';
import Status from './Status';

interface IProps {
  timeToCount: number;
  postProcess?: IPostProcess;
}
interface IState {
  // empty
}
interface IPostProcess {
  onCount?: (remaining: number) => void;
  onCountStart?: () => void;
  onCountStop?: () => void;
  onCountPause?: () => void;
  onCountResume?: () => void;
  onCountComplete?: () => void;
}
class Timer extends React.Component<IProps, IState> {
  // Post process
  private postCount: (remaining: number) => void;
  private postCountStart: () => void;
  private postCountStop: () => void;
  private postCountPause: () => void;
  private postCountResume: () => void;
  private postCountComplete: () => void;

  // Initialize in init()
  private status: Status;
  private timerId!: NodeJS.Timer | number;
  private remaining!: number;

  public constructor(props: IProps) {
    super(props);

    // Post process
    this.postCount = props.postProcess && props.postProcess.onCount ? props.postProcess.onCount : (remaining: number) => {};
    this.postCountStart = props.postProcess && props.postProcess.onCountStart ? props.postProcess.onCountStart : () => {};
    this.postCountStop = props.postProcess && props.postProcess.onCountStop ? props.postProcess.onCountStop : () => {};
    this.postCountPause = props.postProcess && props.postProcess.onCountPause ? props.postProcess.onCountPause : () => {};
    this.postCountResume = props.postProcess && props.postProcess.onCountResume ? props.postProcess.onCountResume : () => {};
    this.postCountComplete = props.postProcess && props.postProcess.onCountComplete ? props.postProcess.onCountComplete : () => {};

    // Method binding
    this.timerRunner = this.timerRunner.bind(this);

    // Local variable
    this.status = Status.Stopped;
    this.remaining = props.timeToCount;
  }
  private isTimerCompleted(remaining: number): boolean {
    return remaining < 0;
  }
  private timerRunner(): void {
    this.remaining--;
    if (this.isTimerCompleted(this.remaining)) {
      this.stop();
      this.postCountComplete();
    } else {
      this.postCount(this.remaining);
    }
  }
  public start(): void {
    this.timerId = setInterval(this.timerRunner, 1000);
    this.postCountStart();
    this.status = Status.Running;
  }
  public stop(): void {
    clearInterval(this.timerId as NodeJS.Timer);
    this.postCountStop();
    this.status = Status.Stopped;
    this.remaining = this.props.timeToCount;
  }
  public pause(): void {
    clearInterval(this.timerId as NodeJS.Timer);
    this.postCountPause();
    this.status = Status.Paused;
  }
  public resume(): void {
    this.timerId = setInterval(this.timerRunner, 1000);
    this.postCountResume();
    this.status = Status.Running;
  }
  public getState(): Status {
    return this.status;
  }
  public getTimeToCount(): number {
    return this.props.timeToCount;
  }
  public setProgress(progress: number): void {
    this.remaining = Math.round(this.props.timeToCount * Utils.safeSubtraction(1, progress));
  }
  public getProgress(): number {
    return Utils.safeSubtraction(1, this.remaining / this.props.timeToCount);
  }
  public toggle(): void {
    switch (this.status) {
        case Status.Running:
        this.pause();
        break;
      case Status.Paused:
        this.resume();
        break;
      case Status.Stopped:
        this.start();
        break;
    }
  }
  /* istanbul ignore next */
  public init(props: IProps): void {
    clearInterval(this.timerId as NodeJS.Timer);
    this.status = Status.Stopped;
    this.remaining = props.timeToCount;
  }
  /* istanbul ignore next */
  public componentWillUnmount(): void {
    clearInterval(this.timerId as NodeJS.Timer);
  }
  /* istanbul ignore next */
  public componentWillReceiveProps(newProps: IProps): void {
    if (_.isEqual(this.props, newProps)) {
      return;
    }
    this.init(newProps);
  }
  public render(): React.ReactNode {
    return null;
  }
}
export default Timer;
