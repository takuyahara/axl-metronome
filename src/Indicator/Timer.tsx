import React, { createRef, RefObject } from 'react';
import _ from 'lodash';
import Utils from '../Utils';
import Status from './Status';

interface IProps {
  timeToCount: number;
  postProcess?: IPostProcess;
}
interface IState {
  // empty
}
interface IPostProcess {
  onCount?: (remaining: number) => void;
  onCountStart?: () => void;
  onCountStop?: () => void;
  onCountPause?: () => void;
  onCountResume?: () => void;
  onCountComplete?: () => void;
}
class Timer extends React.Component<IProps, IState> {
  // Post process
  private postCount: (remaining: number) => void;
  private postCountStart: () => void;
  private postCountStop: () => void;
  private postCountPause: () => void;
  private postCountResume: () => void;
  private postCountComplete: () => void;

  // Initialize in init()
  private status: Status;
  private timerId!: NodeJS.Timer | number;
  private remaining!: number;

  public constructor(props: IProps) {
    super(props);
    this.postCount = props.postProcess && props.postProcess.onCount ? props.postProcess.onCount : (remaining: number) => {};
    this.postCountStart = props.postProcess && props.postProcess.onCountStart ? props.postProcess.onCountStart : () => {};
    this.postCountStop = props.postProcess && props.postProcess.onCountStop ? props.postProcess.onCountStop : () => {};
    this.postCountPause = props.postProcess && props.postProcess.onCountPause ? props.postProcess.onCountPause : () => {};
    this.postCountResume = props.postProcess && props.postProcess.onCountResume ? props.postProcess.onCountResume : () => {};
    this.postCountComplete = props.postProcess && props.postProcess.onCountComplete ? props.postProcess.onCountComplete : () => {};
    this.status = Status.Stopped;

    this.timerRunner = this.timerRunner.bind(this);
    this.init(props);
  }
  private init(props: IProps): void {
    this.remaining = props.timeToCount;
  }
  private isTimerCompleted(remaining: number): boolean {
    return remaining < 0;
  }
  private timerRunner(): void {
    this.remaining--;
    if (this.isTimerCompleted(this.remaining)) {
      this.stop();
      this.postCountComplete();
    } else {
      this.postCount(this.remaining);
    }
  }
  public start(): void {
    this.timerId = setInterval(this.timerRunner, 1000);
    this.postCountStart();
    this.status = Status.Running;
  }
  public stop(): void {
    clearInterval(this.timerId as NodeJS.Timer);
    this.postCountStop();
    this.status = Status.Stopped;
    this.remaining = this.props.timeToCount;
  }
  public pause(): void {
    clearInterval(this.timerId as NodeJS.Timer);
    this.postCountPause();
    this.status = Status.Paused;
  }
  public resume(): void {
    this.timerId = setInterval(this.timerRunner, 1000);
    this.postCountResume();
    this.status = Status.Running;
  }
  public getState(): Status {
    return this.status;
  }
  public getTimeToCount(): number {
    return this.props.timeToCount;
  }
  public setProgress(progress: number): void {
    this.remaining = this.props.timeToCount * (1 - progress);
  }
  public getProgress(): number {
    const progressRaw = 1 - (this.remaining / this.props.timeToCount);
    const progressFixed = Utils.fixNumber(progressRaw); // FIXME: Rounding Error
    return progressFixed;
  }
  public toggle(): void {
    switch (this.status) {
        case Status.Running:
        this.pause();
        break;
      case Status.Paused:
        this.resume();
        break;
      case Status.Stopped:
        this.start();
        break;
    }
  }
  /* istanbul ignore next */
  public componentWillReceiveProps(newProps: IProps): void {
    if (_.isEqual(this.props, newProps)) {
      return;
    }

    this.init(newProps);
  }
  public render(): React.ReactNode {
    return null;
  }
}
export default Timer;
