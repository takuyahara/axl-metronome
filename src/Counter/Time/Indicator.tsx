import React, { Component, createRef, RefObject } from 'react';
import _ from 'lodash';
import Remaining from './Remaining';
import Timer from './Timer';
import Status from './Status';
// import style from './Indicator.module.scss';
import Utils, { ICssSelector } from '../../Utils';

interface IProps {
  timeToCount: number;
  postProcess?: IPostProcess;
  inheritedSelector?: ICssSelector;
}
interface IState {
  timeToCount: number;
}
interface IPostProcess {
  onCount?: (remaining: number) => void;
  onCountStart?: () => void;
  onCountStop?: () => void;
  onCountPause?: () => void;
  onCountResume?: () => void;
  onCountComplete?: () => void;
}
class Indicator extends Component<IProps, IState> {
  // Initialize in init()
  private timer!: Timer;

  // Ref
  private refRemaining: RefObject<Remaining>;
  private refTimer: RefObject<Timer>;

  // Post process
  private postCount: (remaining: number) => void;
  private postCountStart: () => void;
  private postCountStop: () => void;
  private postCountPause: () => void;
  private postCountResume: () => void;
  private postCountComplete: () => void;

  public constructor(props: IProps) {
    super(props);
    this.state = {
      timeToCount: props.timeToCount,
    };

    // Ref
    this.refRemaining = createRef<Remaining>();
    this.refTimer = createRef<Timer>();

    // Binding
    this.onClick = this.onClick.bind(this);
    this.onCount = this.onCount.bind(this);
    this.onCountStart = this.onCountStart.bind(this);
    this.onCountStop = this.onCountStop.bind(this);
    this.onCountPause = this.onCountPause.bind(this);
    this.onCountResume = this.onCountResume.bind(this);
    this.onCountComplete = this.onCountComplete.bind(this);

    // Post process
    this.postCount = props.postProcess && props.postProcess.onCount ? props.postProcess.onCount : (remaining: number) => {};
    this.postCountStart = props.postProcess && props.postProcess.onCountStart ? props.postProcess.onCountStart : () => {};
    this.postCountStop = props.postProcess && props.postProcess.onCountStop ? props.postProcess.onCountStop : () => {};
    this.postCountPause = props.postProcess && props.postProcess.onCountPause ? props.postProcess.onCountPause : () => {};
    this.postCountResume = props.postProcess && props.postProcess.onCountResume ? props.postProcess.onCountResume : () => {};
    this.postCountComplete = props.postProcess && props.postProcess.onCountComplete ? props.postProcess.onCountComplete : () => {};
  }
  private onClick(e: React.MouseEvent | React.TouchEvent): void {
    const timer = this.refTimer.current!;
    timer.toggle();
  }
  private onCount(remaining: number): void {
    const refRemaining = this.refRemaining.current!;
    refRemaining.setRemaining(remaining);
    this.postCount(remaining);
  }
  private onCountStart(): void {
    this.postCountStart();
  }
  private onCountStop(): void {
    this.postCountStop();
  }
  private onCountPause(): void {
    this.postCountPause();
  }
  private onCountResume(): void {
    this.postCountResume();
  }
  private onCountComplete(): void {
    // this.setState({
    //   timeToCount: this.props.timeToCount,
    // });
    const refRemaining = this.refRemaining.current!;
    refRemaining.setRemaining(this.props.timeToCount);
    this.postCountComplete();
  }
  public toggle(): void {
    const timer = this.refTimer.current!;
    timer.toggle();
  }
  public start(): void {
    const timer = this.refTimer.current!;
    timer.start();
  }
  public stop(): void {
    const timer = this.refTimer.current!;
    timer.stop();
  }
  public pause(): void {
    const timer = this.refTimer.current!;
    timer.pause();
  }
  public resume(): void {
    const timer = this.refTimer.current!;
    timer.resume();
  }
  public getTimeToCount(): number {
    const timer = this.refTimer.current!;
    return timer.getTimeToCount();
  }
  public setProgress(newProgress: number): void {
    const refRemaining = this.refRemaining.current!;
    const refTimer = this.refTimer.current!;
    refRemaining.setRemaining(Math.round((1 - newProgress) * this.props.timeToCount));
    refTimer.setProgress(newProgress);
  }
  public getProgress(): number {
    const timer = this.refTimer.current!;
    return timer.getProgress();
  }
  public getState(): Status {
    const timer = this.refTimer.current!;
    return timer.getState();
  }
  /* istanbul ignore next */
  public init(newProps: IProps): void {
    this.stop();
    this.setState({
      timeToCount: newProps.timeToCount,
    });

    const refRemaining = this.refRemaining.current!;
    const refTimer = this.refTimer.current!;
    refRemaining.init({
      timeToCount: newProps.timeToCount,
    });
    refTimer.init({
      timeToCount: newProps.timeToCount,
    });
  }
  // /* istanbul ignore next */
  // public componentWillMount(): void {
  //   if (this.props.inheritedSelector) {
  //     Utils.inheritSelector(style, this.props.inheritedSelector);
  //   }
  // }
  /* istanbul ignore next */
  public componentWillUnmount(): void {
    this.stop();
  }
  /* istanbul ignore next */
  public componentWillReceiveProps(newProps: IProps): void {
    if (_.isEqual(this.props, newProps)) {
      return;
    }
    this.init(newProps);
  }
  public render(): React.ReactNode {
    return (
      <>
        <Remaining 
          ref={this.refRemaining}
          timeToCount={this.state.timeToCount}
          postProcess={{
            onClick: this.onClick,
            onTouchEnd: this.onClick,
          }}
          inheritedSelector={this.props.inheritedSelector} 
          // inheritedSelector={style} 
          />
        <Timer 
          ref={this.refTimer}
          timeToCount={this.state.timeToCount} 
          postProcess={{
            onCount: this.onCount,
            onCountStart: this.onCountStart,
            onCountStop: this.onCountStop,
            onCountPause: this.onCountPause,
            onCountResume: this.onCountResume,
            onCountComplete: this.onCountComplete,
          }} 
          />
      </>
    );
  }
}
export default Indicator;
