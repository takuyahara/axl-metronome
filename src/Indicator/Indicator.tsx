import React, { Component, createRef, RefObject } from 'react';
import _ from 'lodash';
import style from './Indicator.module.scss';
import Timer from './Timer';
import State from './State';

interface IProps {
  timeToCount: number;
  postProcess: IPostProcess;
}
interface IState {
  // elapsedTime: number;
  remaining: number;
}
interface IPostProcess {
  onCount: (remaining: number) => void;
  onCountStart: () => void;
  onCountStop: () => void;
  onCountComplete: (remaining: number) => void;
}
class Indicator extends Component<IProps, IState> {
  // Initialize in init()
  private timer!: Timer;

  // Post process
  private postCount: IPostProcess['onCount'];
  private postCountStart: IPostProcess['onCountStart'];
  private postCountStop: IPostProcess['onCountStop'];
  private postCountComplete: IPostProcess['onCountComplete'];
  public constructor(props: IProps) {
    super(props);
    this.state = {
      remaining: props.timeToCount,
    };

    // Binding
    this.onClick = this.onClick.bind(this);

    // Post process
    this.postCount = props.postProcess.onCount;
    this.postCountStart = props.postProcess.onCountStart;
    this.postCountStop = props.postProcess.onCountStop;
    this.postCountComplete = props.postProcess.onCountComplete;

    // Initialize
    this.init(props);
  }
  private init(props: IProps): void {
    const postProcess = {
      onCount: (remaining: number) => {
        this.setState({
          remaining,
        });
        this.postCount(remaining);
      },
      onCountStart: () => {
        this.postCountStart();
      },
      onCountStop: () => {
        this.postCountStop();
      },
      onCountComplete: (remaining: number) => {
        this.setState({
          remaining,
        });
        this.postCountComplete(remaining);
      },
    };
    this.timer = new Timer(props.timeToCount, {...postProcess});
  }
  private onClick(e: React.MouseEvent | React.TouchEvent): void {
    const isMouseEvent = e.type.substr(0, 5) === 'mouse';
    isMouseEvent && e.preventDefault();
    
    this.timer.toggle();
  }
  private formatTime(timeInSecond: number): string
  {
    const minute = Math.floor(timeInSecond / 60);
    const second = Math.floor(timeInSecond % 60);
    const minuteStr = minute < 10 ? "0" + minute : "" + minute;
    const secondStr = second < 10 ? "0" + second : "" + second;
    return minuteStr + ":" + secondStr;
  }
  public toggle(): void {
    this.timer.toggle();
  }
  public setRemaining(newRemaining: number): void {
    this.timer.setRemaining(newRemaining);
  }
  public getRemaining(): number {
    return this.timer.getRemaining();
  }
  public setProgress(newProgress: number): void {
    this.timer.setProgress(newProgress);
    this.setState({
      remaining: this.timer.getRemaining(),
    });
  }
  public getProgress(): number {
    return this.timer.getProgress();
  }
  public getState(): State {
    return this.timer.getState();
  }
  public reset(): void {
    this.timer.reset();
  }
  public drag(): void {
    this.timer.drag();
  }
  public resume(): void {
    this.timer.resume();
  }
  public componentWillUnmount(): void {
    this.timer.stop();
  }
  public componentWillMount(): void {
    // const postProcess = {
    //   countStart: this.postCountStart,
    //   count: this.postCount,
    //   countStop: this.postCountStop,
    //   countComplete: this.postCountComplete,
    // };
    // this._timer = new Timer(this.props.remaining, this.cbPassRemaining, postProcess);
  }
  public componentWillReceiveProps(newProps: IProps): void {
    if (_.isEqual(this.props, newProps)) {
      return;
    }

    this.timer.stop();

    this.init(newProps);
    this.setState({
      remaining: newProps.timeToCount,
    });
  }
  public render(): React.ReactNode {
    const formattedTime = this.formatTime(Math.round(this.state.remaining));
    return (
      <div className={style.indicator} onClick={this.onClick}>
        <div className={style.remaining}>
          {formattedTime}
        </div>
      </div>
    );
  }
}
export default Indicator;
