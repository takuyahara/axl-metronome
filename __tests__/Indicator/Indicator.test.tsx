import React from 'react';
import { configure, mount, ReactWrapper } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import sinon from 'sinon';
import Indicator from 'Indicator/Indicator';
import Timer from 'Indicator/Timer';
import State from 'Indicator/State';

configure({ adapter: new Adapter() });

describe('Indicator', () => {
  let indicator: ReactWrapper;
  let indicatorInstance: Indicator;
  let spy: {
    onClick: sinon.SinonSpy,
    toggle: sinon.SinonSpy,
  };
  let clock: sinon.SinonFakeTimers;
  const postProcess = {
    onCount: (remaining: number) => {},
    onCountStart: () => {},
    onCountStop: () => {},
    onCountComplete: (remaining: number) => {},
  };
  
  beforeEach(() => {
    spy = Object.assign({
      onClick: sinon.spy((Indicator as any).prototype, 'onClick'),
      toggle: sinon.spy(Timer.prototype, 'toggle'),
    });
    clock = sinon.useFakeTimers();
    indicator = mount(<Indicator 
      timeToCount={900} 
      postProcess={{...postProcess}} 
    />);
    indicatorInstance = indicator.instance() as Indicator;
  })
  afterEach(() => {
    sinon.restore();
  });

  describe('public functions', () => {
    it('changes remaining time', () => {
      indicatorInstance.setProgress(0.25);
      expect(indicatorInstance.getProgress()).toBe(0.25);
      expect(indicatorInstance.getRemaining()).toBe(675);
    });

    it('calls reset when not running', () => {
      expect(indicatorInstance.getState()).toBe(State.Stopped);
      expect(indicatorInstance.getRemaining()).toBe(900);
      (indicatorInstance as any).timer.reset();
      expect(indicatorInstance.getState()).toBe(State.Stopped);
      expect(indicatorInstance.getRemaining()).toBe(900);
    });

    it('calls reset when running', () => {
      indicator.find('.remaining').simulate('click', {
        clientX: 70,
        clientY: 70,
      });
      clock.tick(2000);

      expect(indicatorInstance.getState()).toBe(State.Running);
      expect(indicatorInstance.getRemaining()).toBe(898);

      indicatorInstance.reset();
      clock.tick(2000);
      expect(indicatorInstance.getState()).toBe(State.Stopped);
      expect(indicatorInstance.getRemaining()).toBe(900);
    });
  });

  describe('private functions', () => {
    it('formats number', () => {
      expect((indicatorInstance as any).formatTime(0)).toBe('00:00');
      expect((indicatorInstance as any).formatTime(9)).toBe('00:09');
      expect((indicatorInstance as any).formatTime(10)).toBe('00:10');
      expect((indicatorInstance as any).formatTime(59)).toBe('00:59');
      expect((indicatorInstance as any).formatTime(60)).toBe('01:00');
      expect((indicatorInstance as any).formatTime(69)).toBe('01:09');
      expect((indicatorInstance as any).formatTime(70)).toBe('01:10');
      expect((indicatorInstance as any).formatTime(600)).toBe('10:00');
      expect((indicatorInstance as any).formatTime(3599)).toBe('59:59');
    });
  });

  describe('MouseEvents', () => {
    afterEach(() => {
      indicator.update();
      expect(indicator).toMatchSnapshot();
    });
    it('clicks 1 time', () => {
      indicator.find('.remaining').simulate('click', {
        clientX: 70,
        clientY: 70,
      });
      expect(spy.onClick.calledOnce).toBe(true);
      expect(spy.toggle.calledOnce).toBe(true);
      clock.tick(2000);

      expect(indicatorInstance.getState()).toBe(State.Running);
      expect(indicatorInstance.getRemaining()).toBe(898);
    });

    it('clicks 2 times', () => {
      // First click
      indicator.find('.remaining').simulate('click', {
        clientX: 70,
        clientY: 70,
      });
      expect(spy.onClick.calledOnce).toBe(true);
      expect(spy.toggle.calledOnce).toBe(true);

      // First wait
      clock.tick(2000);
      expect(indicatorInstance.getState()).toBe(State.Running);
      expect(indicatorInstance.getRemaining()).toBe(898);

      // Second click
      spy.onClick.resetHistory();
      spy.toggle.resetHistory();
      indicator.find('.remaining').simulate('click', {
        clientX: 70,
        clientY: 70,
      });
      expect(spy.onClick.calledOnce).toBe(true);
      expect(spy.toggle.calledOnce).toBe(true);

      // Second wait
      clock.tick(2000);
      expect(indicatorInstance.getState()).toBe(State.Paused);
      expect(indicatorInstance.getRemaining()).toBe(898);

      // Done
    });

    it('clicks 3 times', () => {
      // First click
      indicator.find('.remaining').simulate('click', {
        clientX: 70,
        clientY: 70,
      });
      expect(spy.onClick.calledOnce).toBe(true);
      expect(spy.toggle.calledOnce).toBe(true);

      // First wait
      clock.tick(2000);
      expect(indicatorInstance.getState()).toBe(State.Running);
      expect(indicatorInstance.getRemaining()).toBe(898);

      // Second click
      spy.onClick.resetHistory();
      spy.toggle.resetHistory();
      indicator.find('.remaining').simulate('click', {
        clientX: 70,
        clientY: 70,
      });
      expect(spy.onClick.calledOnce).toBe(true);
      expect(spy.toggle.calledOnce).toBe(true);

      // Second wait
      clock.tick(2000);
      expect(indicatorInstance.getState()).toBe(State.Paused);
      expect(indicatorInstance.getRemaining()).toBe(898);

      // Third click
      spy.onClick.resetHistory();
      spy.toggle.resetHistory();
      indicator.find('.remaining').simulate('click', {
        clientX: 70,
        clientY: 70,
      });
      expect(spy.onClick.calledOnce).toBe(true);
      expect(spy.toggle.calledOnce).toBe(true);

      // Third wait
      clock.tick(2000);
      expect(indicatorInstance.getState()).toBe(State.Running);
      expect(indicatorInstance.getRemaining()).toBe(896);
    });
  });
});
