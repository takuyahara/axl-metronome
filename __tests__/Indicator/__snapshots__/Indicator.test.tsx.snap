// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indicator MouseEvents clicks 1 time 1`] = `
<Indicator
  postProcess={
    Object {
      "onCount": [Function],
      "onCountComplete": [Function],
      "onCountStart": [Function],
      "onCountStop": [Function],
    }
  }
  timeToCount={900}
>
  <div
    className="indicator"
    onClick={[Function]}
  >
    <div
      className="remaining"
    >
      14:58
    </div>
  </div>
</Indicator>
`;

exports[`Indicator MouseEvents clicks 2 times 1`] = `
<Indicator
  postProcess={
    Object {
      "onCount": [Function],
      "onCountComplete": [Function],
      "onCountStart": [Function],
      "onCountStop": [Function],
    }
  }
  timeToCount={900}
>
  <div
    className="indicator"
    onClick={[Function]}
  >
    <div
      className="remaining"
    >
      14:58
    </div>
  </div>
</Indicator>
`;

exports[`Indicator MouseEvents clicks 3 times 1`] = `
<Indicator
  postProcess={
    Object {
      "onCount": [Function],
      "onCountComplete": [Function],
      "onCountStart": [Function],
      "onCountStop": [Function],
    }
  }
  timeToCount={900}
>
  <div
    className="indicator"
    onClick={[Function]}
  >
    <div
      className="remaining"
    >
      14:56
    </div>
  </div>
</Indicator>
`;
